File: python3/intersection_of_two_linked_lists.py

from .model import ListNode

"""
Space   : O(1)
Time    : O(n)
"""


class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        na, nb = 0, 0
        pa, pb = headA, headB

        while pa:
            pa = pa.next
            na += 1

        while pb:
            pb = pb.next
            nb += 1

        pa, pb = headA, headB

        while na > nb:
            pa = pa.next
            na -= 1

        while nb > na:
            pb = pb.next
            nb -= 1

        while pa and pb:
            if pa == pb:
                return pa
            pa = pa.next
            pb = pb.next

        return None

File: golang/intersection_of_two_linked_lists.go
package main

import (
	"golang/model"
)

// Space	: O(1)
// Time	: O(n)

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func getIntersectionNode(headA, headB *model.ListNode) *model.ListNode {
	la, lb := 0, 0
	pa, pb := headA, headB

	for pa != nil {
		la++
		pa = pa.Next
	}

	for pb != nil {
		lb++
		pb = pb.Next
	}

	pa, pb = headA, headB

	dif := 0
	if la > lb {
		dif = la - lb
		for dif > 0 {
			pa = pa.Next
			dif--
		}
	} else {
		dif = lb - la
		for dif > 0 {
			pb = pb.Next
			dif--
		}
	}

	for pa != nil {
		if pa == pb {
			return pa
		}
		pa = pa.Next
		pb = pb.Next
	}

	return nil

}
